From 33bea44922713dc2355f03e5f71d07bb2d6147e9 Mon Sep 17 00:00:00 2001
From: Omri Itach <omrii@marvell.com>
Date: Thu, 13 Jun 2013 14:25:03 +0300
Subject: [PATCH 0722/1825] ALP: Fixed mvCtrlBoardConfigGet: Added WA for IO
 Exp. 0x21 bug (DB-6660 board)

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit 6fad4e51b7d97542c0c4e9e7fc909a58cae84d69

- Bug: Pins at IO expander 0x21 are reversed (only on DB-6660)
  Solution: reverse bits after reading IO expander
- Fixed read of 2nd Dip Switch reg at IO expander 0x21
- Fixed read of 3rd DIP switch reg at IO expander 0x24 (for DB-6660)

Change-Id: Ib4f46a7b9554f0111b42407587588a5b7c67b252
Signed-off-by: Omri Itach <omrii@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/2239
Reviewed-by: Eran Ben-Avi <benavi@marvell.com>
Tested-by: Eran Ben-Avi <benavi@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .../avanta_lp_family/ctrlEnv/mvCtrlEnvLib.c        |   12 ++++++++++--
 1 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/arch/arm/mach-avantalp/avanta_lp_family/ctrlEnv/mvCtrlEnvLib.c b/arch/arm/mach-avantalp/avanta_lp_family/ctrlEnv/mvCtrlEnvLib.c
index 561df60..25d709e 100755
--- a/arch/arm/mach-avantalp/avanta_lp_family/ctrlEnv/mvCtrlEnvLib.c
+++ b/arch/arm/mach-avantalp/avanta_lp_family/ctrlEnv/mvCtrlEnvLib.c
@@ -625,7 +625,7 @@ MV_STATUS mvCtrlBoardConfigGet(MV_U8 *config)
 	MV_BOARD_TWSI_CLASS twsiClass = (isEepromEnabled ? BOARD_DEV_TWSI_EEPROM : BOARD_DEV_TWSI_IO_EXPANDER);
 
 	status1 = mvBoardTwsiGet(twsiClass, 0, 0, &config[0]);		/* EEPROM/Dip Switch Reg#0 */
-	status2 = mvBoardTwsiGet(twsiClass, 0, 0, &config[1]);		/* EEPROM/Dip Switch Reg#1 */
+	status2 = mvBoardTwsiGet(twsiClass, 0, 1, &config[1]);		/* EEPROM/Dip Switch Reg#1 */
 
 	if (status1 != MV_OK || status2 != MV_OK) {
 		DB(mvOsPrintf("%s: Error: mvBoardTwsiGet from EEPROM/Dip Switch failed\n", __func__));
@@ -633,7 +633,15 @@ MV_STATUS mvCtrlBoardConfigGet(MV_U8 *config)
 	}
 
 	if (boardId == DB_6660_ID) { /* DB-6660 has another register for board configuration */
-		if (isEepromEnabled == MV_OK)
+		/*
+		 * Workaround for DIP Switch IO Expander 0x21 bug in DB-6660 board
+		 * Bug: Pins at IO expander 0x21 are reversed (only on DB-6660)
+		 * Solution: reverse bits after reading IO expander
+		 */
+		config[0] = mvReverseBits(config[0]);
+		config[1] = mvReverseBits(config[1]);
+
+		if (isEepromEnabled == MV_TRUE)
 			status1 = mvBoardTwsiGet(BOARD_DEV_TWSI_EEPROM, 0, 2, &config[2]);	/* EEPROM Reg#2 */
 		else
 			status1 = mvBoardTwsiGet(BOARD_DEV_TWSI_IO_EXPANDER, 1, 0, &config[2]);	/* Dip Switch Reg#1 */
-- 
1.7.5.4

